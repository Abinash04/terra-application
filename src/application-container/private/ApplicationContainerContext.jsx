import { createContext } from 'react';

const ApplicationContainerContext = createContext({
  disableParentFill: false,
});

export default ApplicationContainerContext;
