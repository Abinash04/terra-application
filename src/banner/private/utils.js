const BANNER_TYPES = {
  ALERT: 'alert',
  ERROR: 'error',
  WARNING: 'warning',
  UNSATISFIED: 'unsatisfied',
  UNVERIFIED: 'unverified',
  ADVISORY: 'advisory',
  INFO: 'info',
  SUCCESS: 'success',
  CUSTOM: 'custom',
};

const defaultBannerPriority = [
  BANNER_TYPES.ALERT,
  BANNER_TYPES.ERROR,
  BANNER_TYPES.WARNING,
  BANNER_TYPES.UNSATISFIED,
  BANNER_TYPES.UNVERIFIED,
  BANNER_TYPES.ADVISORY,
  BANNER_TYPES.INFO,
  BANNER_TYPES.SUCCESS,
  BANNER_TYPES.CUSTOM,
];

const organizeBannersByPriority = (banners) => {
  if (!banners) {
    return [];
  }

  const bannersList = defaultBannerPriority.reduce((coll, type) => {
    if (banners[type]) {
      coll.push(type);
    }
    return coll;
  }, []);

  return bannersList;
};

export { organizeBannersByPriority, BANNER_TYPES };
