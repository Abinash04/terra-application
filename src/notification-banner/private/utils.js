const BANNER_VARIANTS = {
  HAZARD_HIGH: 'hazard-high',
  HAZARD_MEDIUM: 'hazard-medium',
  HAZARD_LOW: 'hazard-low',
  ERROR: 'error',
  UNSATISFIED: 'unsatisfied',
  UNVERIFIED: 'unverified',
};

const defaultBannerPriority = [
  BANNER_VARIANTS.HAZARD_HIGH,
  BANNER_VARIANTS.HAZARD_MEDIUM,
  BANNER_VARIANTS.HAZARD_LOW,
  BANNER_VARIANTS.ERROR,
  BANNER_VARIANTS.UNSATISFIED,
  BANNER_VARIANTS.UNVERIFIED,
];

const fusionBannerPriority = [
  BANNER_VARIANTS.HAZARD_HIGH,
  BANNER_VARIANTS.ERROR,
  BANNER_VARIANTS.HAZARD_MEDIUM,
  BANNER_VARIANTS.UNSATISFIED,
  BANNER_VARIANTS.UNVERIFIED,
  BANNER_VARIANTS.HAZARD_LOW,
];

/**
 * Organize / prioritized the banners for the given theme.
 * @param {Object[]} banners - the list of banner props to organize
 * @param {String} themeName - name of the theme to prioritize the banners for. <---- TO DO - verify this updates correctly
 *
 * @returns List of prioritized banners.
 */
const organizeBannersByPriority = (banners, themeName) => {
  if (!banners) {
    return [];
  }

  let priorityOrder = defaultBannerPriority;
  if (themeName === 'orion-fusion-theme') {
    priorityOrder = fusionBannerPriority;
  }

  const bannersList = priorityOrder.reduce((coll, type) => {
    if (banners[type]) {
      // eslint-disable-next-line compat/compat
      const registeredBannersForType = Object.values(banners[type]);
      registeredBannersForType.map(registeredBannerProps => coll.push(registeredBannerProps));
    }

    return coll;
  }, []);

  return bannersList;
};

export { organizeBannersByPriority, BANNER_VARIANTS };
