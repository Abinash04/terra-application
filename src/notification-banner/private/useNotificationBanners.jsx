import React from 'react';
import PropTypes from 'prop-types';
import Alert from 'terra-alert';
import Button from 'terra-button';
import BannerRegistrationContext from './BannerRegistrationContext';

import { organizeBannersByPriority } from './utils';

/**
 * The `useNotificationBanners` Hook manages registering and prioritizing Notification Banners
 * rendered within the Notification Banner Context.
 *
 * Returns:
 *   - NotificationBannerProvider - React Context Provider - Provides the Banner Registration Context to its children.
 *         This allows any NotificationBanner registered beneath it to be displayed in teh NotificationBanner's list.
 *   - NotificationBanners - React Component - Renders a list of prioritized notification banners.
 *
 * @returns { NotificationBannerProvider, NotificationBanners }
 */
const useNotificationBanners = () => {
  /** The registeredBanners ref maintains collection of banner ids and banner props registered to the  NotificationBannerProvider. */
  const registeredBanners = React.useRef({});

  /**
   * The updateBannerState ref stores the update state function used to manage the banner rendered in the NotificationBanners component.
   * This ties the state updates to the `useNotificationBanners` hook, while allowing the NotificationBanners to be rendered above or below
   * the NotificationBannerProvider.
   */
  const updateBannerState = React.useRef();

  const bannerProviderValue = React.useMemo(() => {
    /**
     * Adds the banner ID and props to the collection of registered banners. Once registered,
     * it updates the NotificationBanner's state to render the new Banner.
     *
     * @param {UUID} bannerId - unique ID associated to the banner
     * @param {Object} bannerProps - react props associated tot he banner. See ./NotificationBanner's propTypes.
     */
    const registerNotificationBanner = (bannerId, bannerProps) => {
      if (process.env.NODE_ENV !== 'production' && !bannerId) {
        // eslint-disable-next-line no-console
        console.warn('A banner cannot be registered without an identifier.');
        return;
      }

      const { type } = bannerProps;

      if (!registeredBanners.current[type]) {
        registeredBanners.current[type] = {};
      }

      registeredBanners.current[type][bannerId] = { key: bannerId, ...bannerProps };

      if (updateBannerState.current) {
        updateBannerState.current(organizeBannersByPriority(registeredBanners.current));
      }
    };

    /**
     * Removes the banner ID and props from the collection of registered banners. Once unregistered,
     * it updates the NotificationBanner's state to remove the Banner from the list.
     *
     * @param {UUID} bannerId - unique ID associated to the banner
     * @param {String} bannerType - the banner variant
     */
    const unregisterNotificationBanner = (bannerId, bannerType) => {
      if (process.env.NODE_ENV !== 'production' && (!bannerId || !bannerType)) {
        // eslint-disable-next-line no-console
        console.warn('A banner cannot be unregistered without an identifier or banner type.');
        return;
      }

      if (!registeredBanners.current[bannerType][bannerId]) {
        return;
      }

      delete registeredBanners.current[bannerType][bannerId];

      if (updateBannerState.current) {
        updateBannerState.current(organizeBannersByPriority(registeredBanners.current));
      }
    };

    return {
      registerNotificationBanner,
      unregisterNotificationBanner,
    };
  }, []);

  /**
   * Provides the Banner Registration Context to its children.
   */
  const NotificationBannerProvider = ({ children }) => (
    <BannerRegistrationContext.Provider value={bannerProviderValue}>
      {children}
    </BannerRegistrationContext.Provider>
  );

  NotificationBannerProvider.propTypes = {
    children: PropTypes.node,
  };

  /**
   * Renders a list of prioritized notification banners.
   */
  const NotificationBanners = () => {
    const [banners, setBanners] = React.useState([]);
    updateBannerState.current = setBanners;

    if (!banners.length) {
      return null;
    }

    return (
      <div aria-live="polite">
        {banners.map((bannerProps) => {
          const {
            description, type, bannerAction, onRequestClose, key,
          } = bannerProps;

          let actionButton = null;
          if (bannerAction) {
            actionButton = (
              <Button
                text={bannerAction.text}
                variant="ghost"
                onClick={bannerAction.onClick}
              />
            );
          }

          return (
            <Alert
              key={key}
              action={actionButton}
              onDismiss={onRequestClose}
              type={type}
              data-terra-application-notification-banner={type}
            >
              {description}
            </Alert>
          );
        })}
      </div>
    );
  };

  const useNotificationBannerExports = React.useRef({
    NotificationBannerProvider,
    NotificationBanners,
  });

  return useNotificationBannerExports.current;
};

export default useNotificationBanners;
