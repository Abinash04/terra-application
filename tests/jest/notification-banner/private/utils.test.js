import { BANNER_TYPES, organizeBannersByPriority } from '../../../../src/notification-banner/private/utils';

const alertBannerProps = { id: BANNER_TYPES.ALERT, type: 'alert' };
const errorBannerProps = { id: BANNER_TYPES.ERROR, type: 'error' };
const warningBannerProps = { id: BANNER_TYPES.WARNING, type: 'warning' };
const unsatisfiedBannerProps = { id: BANNER_TYPES.UNSATISFIED, type: 'unsatisfied' };
const unverifiedBannerProps = { id: BANNER_TYPES.UNVERIFIED, type: 'unverified' };
const advisoryBannerProps = { id: BANNER_TYPES.ADVISORY, type: 'advisory' };

describe('utils', () => {
  describe('organizes banner by priority', () => {
    it('returns empty array when banners is undefined', () => {
      const organizedBanners = organizeBannersByPriority();
      expect(organizedBanners).toEqual([]);
    });

    it('returns list of banner props', () => {
      const banners = {
        alert: {
          'randomID-1': alertBannerProps,
        },
        error: {
          'randomID-2': errorBannerProps,
        },
      };

      const organizedBanners = organizeBannersByPriority(banners);
      expect(organizedBanners).toHaveLength(2);
      expect(organizedBanners[0]).toEqual(expect.objectContaining(alertBannerProps));
      expect(organizedBanners[1]).toEqual(expect.objectContaining(errorBannerProps));
    });

    it('organizes banner by priority', () => {
      const banners = {
        unsatisfied: { 'randomID-1': unsatisfiedBannerProps },
        error: { 'randomID-1324': errorBannerProps },
        warning: { 'randomID-1234': warningBannerProps },
        alert: { 'randomID-1333': alertBannerProps },
        unverified: { 'randomID-9081': unverifiedBannerProps },
        advisory: { 'randomID-981': advisoryBannerProps },
      };

      const organizedBanners = organizeBannersByPriority(banners);
      expect(organizedBanners).toMatchSnapshot();

      expect(organizedBanners).toHaveLength(6);
      expect(organizedBanners[0].id).toEqual(BANNER_TYPES.ALERT);
      expect(organizedBanners[1].id).toEqual(BANNER_TYPES.ERROR);
      expect(organizedBanners[2].id).toEqual(BANNER_TYPES.WARNING);
      expect(organizedBanners[3].id).toEqual(BANNER_TYPES.UNSATISFIED);
      expect(organizedBanners[4].id).toEqual(BANNER_TYPES.UNVERIFIED);
      expect(organizedBanners[5].id).toEqual(BANNER_TYPES.ADVISORY);
    });

    it('allows multiple banners by of one type', () => {
      const banners = {
        warning: {
          'randomID-981': warningBannerProps,
          'randomID-92081': { type: 'warning', id: 'second-banner-of-type-warning' },
        },
      };

      const organizedBanners = organizeBannersByPriority(banners);
      expect(organizedBanners).toMatchSnapshot();

      expect(organizedBanners).toHaveLength(2);
      expect(organizedBanners[0].id).toEqual(BANNER_TYPES.WARNING);
      expect(organizedBanners[1].id).toEqual('second-banner-of-type-warning');
    });
  });
});
