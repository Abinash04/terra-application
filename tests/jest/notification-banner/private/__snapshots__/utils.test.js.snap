// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils organizes banner by priority allows multiple banners by of one type 1`] = `
Array [
  Object {
    "id": "warning",
    "type": "warning",
  },
  Object {
    "id": "second-banner-of-type-warning",
    "type": "warning",
  },
]
`;

exports[`utils organizes banner by priority organizes banner by priority 1`] = `
Array [
  Object {
    "id": "alert",
    "type": "alert",
  },
  Object {
    "id": "error",
    "type": "error",
  },
  Object {
    "id": "warning",
    "type": "warning",
  },
  Object {
    "id": "unsatisfied",
    "type": "unsatisfied",
  },
  Object {
    "id": "unverified",
    "type": "unverified",
  },
  Object {
    "id": "advisory",
    "type": "advisory",
  },
]
`;
